### PR TITLE
Added flows:

### PR DESCRIPTION
**Pull request by roykreiner-s**
1. close room option
2. delete room option
3. export room option (exports the playlist to spotify library)

- removed users list action button from settings

### CODE REVIEW

File: app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java [hunk 1]
```diff
@@ -154,31 +163,125 @@ private void initRecyclerViewWithExistingRooms(List<Room> rooms) {
         IRoomActionListener actionListener = new IRoomActionListener() { //TODO implement
             @Override
             public void onRoomExport(Room room) {
-                Toast.makeText(ExistingRoomsActivity.this, "export clicked!", Toast.LENGTH_SHORT).show();
+                HashMap<String, Object> bodyDetails = new HashMap<>(1);
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java:166]: Use generic Map:
`Map<String, Object> body = new HashMap<>();`

File: app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java [hunk 2]
```diff
@@ -154,31 +163,125 @@ private void initRecyclerViewWithExistingRooms(List<Room> rooms) {
...
+                        .followPlaylist(room.getPlaylist().getId(), bodyDetails)
+                        .enqueue(new Callback<Void>() {
+                            @Override
+                            public void onResponse(@NotNull Call<Void> call, @NotNull Response<Void> response) {
+                                Toast.makeText(ExistingRoomsActivity.this, "You can now find the playlist on spotify library", Toast.LENGTH_SHORT).show();
+                            }
+
+                            @Override
+                            public void onFailure(@NotNull Call<Void> call, @NotNull Throwable t) {
+                                Log.e("ExistingRoomsActivity", t.getMessage(), t);
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java:179]: Maybe add a toast "Error saving playlist"?

File: app/src/main/java/com/postpc/tenq/network/SpotifyApiService.java [hunk 3]
```diff
@@ -48,7 +48,13 @@
     @PUT("playlists/{playlist_id}/tracks")
     Call<Void> reorderPlaylist(@Path("playlist_id") String playlistId, @Body() Map<String, Object> details);
 
+    @PUT("playlists/{playlist_id}/followers")
+    Call<Void> followPlaylist(@Path("playlist_id") String playlistId, @Body() Map<String, Object> details);
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/network/SpotifyApiService.java:52]: I'm thinking about the flow of room for guests.... I think when joining a room, this api should be called, no?
And when clicking "export", a new identical playlist should be created.
If the user just follows, the playlist might change if the host manually changes it in his Spotify account. What do you think?

File: app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java [hunk 4]
```diff
@@ -154,31 +163,125 @@ private void initRecyclerViewWithExistingRooms(List<Room> rooms) {
...
             public void onRoomDelete(Room room) {
                 Toast.makeText(ExistingRoomsActivity.this, "delete clicked!", Toast.LENGTH_SHORT).show();
+                User curUser = getAuthService().getCurrentUser();
+                deleteRoomForUser(room.getId(), curUser);
             }
 
             @Override
             public void onRoomClose(Room room) {
-                Toast.makeText(ExistingRoomsActivity.this, "close clicked!", Toast.LENGTH_SHORT).show();
+                if (!room.getHost().getId().equals(getAuthService().getCurrentUser().getId())){
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java:194]: Let's hide the "close" button for guests, so it's not even an option? I think it's a better experience :)

File: app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java [hunk 5]
```diff
@@ -154,31 +163,125 @@ private void initRecyclerViewWithExistingRooms(List<Room> rooms) {
...
+                    Toast.makeText(ExistingRoomsActivity.this, "Only room host can close room", Toast.LENGTH_SHORT).show();
+                }
+                else if (!room.isActive()) {
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java:197]: Same here, the "close" button should be hidden if the room is not active

File: app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java [hunk 6]
```diff
@@ -154,31 +163,125 @@ private void initRecyclerViewWithExistingRooms(List<Room> rooms) {
...
+                    Toast.makeText(ExistingRoomsActivity.this, "room already closed", Toast.LENGTH_SHORT).show();
+                    return;
+                } else {
+                    LayoutInflater inflater = (LayoutInflater) ExistingRoomsActivity.this.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+                    View view = inflater.inflate(R.layout.dialog_close_room, null);
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java:202]: Why do you need a custom view for the dialog?
Use the dialog builder "setMessage" and "setTitle" functions:
image

File: app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java [hunk 7]
```diff
@@ -32,7 +40,8 @@
 import retrofit2.Response;
 
 public class ExistingRoomsActivity extends TenQActivity {
-
+    private final MutableLiveData<ViewModelError> error = new MutableLiveData<>(null);
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/ui/activities/ExistingRoomsActivity.java:43]: Please move these and all network communication to a viewmodel class, like in JoinActivityViewModel

File: app/src/main/java/com/postpc/tenq/ui/activities/RoomActivity.java [hunk 8]
```diff
@@ -54,6 +54,7 @@
     private ActivityRoomBinding binding;
     private ItemTouchHelper itemTouchHelper;
     private Room room;
+    private TracksAdapter adapter;
```
Comment (NitsanAmit) [inline app/src/main/java/com/postpc/tenq/ui/activities/RoomActivity.java:57]: Why?

**Review by NitsanAmit — DISMISSED** [review]
Nice!
Let me know if you wanna talk about the follow/export flow.
Right now the playlist is created when the room is created, and it is public and collaborative. I think that **guests should automatically follow the playlist when they join the room, and "unfollow" when they leave.**
**Exporting** should create a new, identical playlist, private only to the user who exported it - so it sort of "freezes" the room playlist at that moment and creates a private copy of it.

**Comment by roykreiner-s** [comment]
Megre.

**Review by NitsanAmit — APPROVED** [review]

**Pull request merged**